lib/teiserver/game/servers/balancer_server.ex

```diff
@@ -28,34 +27,75 @@ defmodule Teiserver.Game.BalancerServer do
           mean_diff_max: state.mean_diff_max,
           stddev_diff_max: state.stddev_diff_max,
           fuzz_multiplier: state.fuzz_multiplier,
-          shuffle_first_pick: state.shuffle_first_pick
+          shuffle_first_pick: state.shuffle_first_pick,
+          last_balance_hash_cache_hit: state.last_balance_hash_cache_hit,
+          last_balance_hash_cache_miss: state.last_balance_hash_cache_miss,
+          last_balance_hash: state.last_balance_hash,
+          last_balance_result: state.last_balance_result
         ]
 
     {balance, new_state} = make_balance(team_count, state, opts)
     {:reply, balance, new_state}
   end
 
+  @impl true
+  # http://planetspads.free.fr/spads/doc/spadsPluginApiDoc.html#balanceBattle-self-players-bots-clanMode-nbTeams-teamSize
+  def handle_call({:make_balance, team_count, call_opts, players}, _from, state) do
+    opts =
+      call_opts ++
+        [
+          algorithm: state.algorithm,
+          max_deviation: state.max_deviation,
+          rating_lower_boundary: state.rating_lower_boundary,
+          rating_upper_boundary: state.rating_upper_boundary,
+          mean_diff_max: state.mean_diff_max,
+          stddev_diff_max: state.stddev_diff_max,
+          fuzz_multiplier: state.fuzz_multiplier,
+          shuffle_first_pick: state.shuffle_first_pick,
+          last_balance_hash_cache_hit: state.last_balance_hash_cache_hit,
+          last_balance_hash_cache_miss: state.last_balance_hash_cache_miss,
+          last_balance_hash: state.last_balance_hash,
+          last_balance_result: state.last_balance_result
+        ]
+
+    {balance, new_state} = make_balance(team_count, state, opts, players)
+    {:reply, balance, new_state}
+  end
+
   def handle_call(:get_balance_mode, _from, %{last_balance_hash: hash} = state) do
-    result = state.hashes[hash]
+    result =
+      cond do
+        state.last_balance_hash == hash -> state.last_balance_result
+        true -> nil
+      end
+
     {:reply, result.balance_mode, state}
   end
 
   def handle_call(:get_current_balance, _from, %{last_balance_hash: hash} = state) do
-    result = state.hashes[hash]
+    result =
+      cond do
+        state.last_balance_hash == hash -> state.last_balance_result
+        true -> nil
+      end
+
     {:reply, result, state}
   end
 
   def handle_call(:report_state, _from, state) do
     result = %{
-      hashes: Enum.count(state.hashes),
```

Comment:
Oh ... Technically this becomes an API change. I guess I'll check if anything acts on the number of hashes? Does anything do any reporting based on that?

Reply:
Looks like nothing inside Teiserver acts on the number of hashes, so I don't think this will be a breaking change. If desired, I can return a fixed value.